cal.com/generate-url-patterns/index.ts

```diff
@@ -0,0 +1,402 @@
+const OPTIONAL_CATCH_ALL_DYNAMIC_SEGMENT_PART_REGEX =
+	/^\[{2}\.{3}(?<ocadspart>[a-zA-Z0-9]+)\]{2}$/;
+
+const isNeitherNullNorUndefined = <T>(
+	t: NonNullable<T> | null | undefined,
+): t is NonNullable<T> => t !== null && t !== undefined;
+
+const getRegexGroups = (part: string) => {
+	const regExpExecArray =
+		COMMON_PART_REGEX.exec(part) ??
+		DYNAMIC_SEGMENT_PART_REGEX.exec(part) ??
+		CATCH_ALL_DYNAMIC_SEGMENT_PART_REGEX.exec(part) ??
+		OPTIONAL_CATCH_ALL_DYNAMIC_SEGMENT_PART_REGEX.exec(part);
+
+	return regExpExecArray?.groups ?? {};
+};
+
+const handleFile: HandleFile<Dependencies, State> = async (
+	api,
+	path,
+	options,
+	state,
+) => {
+	if (state === null) {
+		throw new Error('The state is not set');
+	}
+
+	if (state.step === 'READING') {
+		const parsedPath = posix.parse(path);
+		const directoryNames = parsedPath.dir.split(posix.sep);
+
+		if (!directoryNames.includes('pages')) {
+			return [];
+		}
+
+		const parts = directoryNames
+			.slice(directoryNames.lastIndexOf('pages') + 1)
+			.concat(parsedPath.name === 'index' ? [] : [parsedPath.name]);
+
+		const pathname = parts
+			.map((part) => {
+				const { cpart, dspart, cadspart, ocadspart } =
+					getRegexGroups(part);
+
+				if (cpart !== undefined) {
+					return cpart;
+				}
+
+				if (dspart !== undefined) {
+					return `:${dspart}`;
+				}
+
+				if (cadspart !== undefined) {
+					return `:${cadspart}+`;
+				}
+
+				if (ocadspart !== undefined) {
+					return `:${ocadspart}*`;
+				}
+
+				return null;
+			})
+			.filter(isNeitherNullNorUndefined)
+			.join('/');
+
+		const partialEnvVar = parts
+			.map((part) => {
+				const { cpart, dspart, cadspart, ocadspart } =
+					getRegexGroups(part);
+
+				const somePart =
+					cpart ?? dspart ?? cadspart ?? ocadspart ?? null;
+
+				return somePart?.toUpperCase() ?? null;
+			})
+			.filter(isNeitherNullNorUndefined)
+			.join('_');
+
+		const envVar = ['APP_ROUTER', partialEnvVar, 'ENABLED'].join('_');
+
+		state.entries.push({
+			pathname,
+			envVar,
+		});
+
+		return [];
+	}
+
+	if (path === state.turboPath || path === state.middlewarePath) {
+		return [
+			{
+				kind: 'upsertFile',
+				path,
+				options,
+			},
+		];
+	}
+
+	return [];
+};
+
+const buildRoutes = (entries: ReadonlyArray<Entry>) => {
+	const statements = entries
+		.map(
+			({ pathname, envVar }) =>
+				`["${pathname}", Boolean(process.env.${envVar})] as const,`,
+		)
+		.join('\n');
+
+	return `
+	const ROUTES: [URLPattern, boolean][] = [
+		${statements}
+  	].map(([pathname, enabled]) => [new URLPattern({ pathname }), enabled]);
+	`;
+};
+
+const handleData: HandleData<Dependencies, State> = async (
+	api,
+	path,
+	data,
+	options,
+	state,
+) => {
+	if (state === null) {
+		throw new Error('The state is not set');
+	}
+
+	if (path === state.turboPath) {
+		const json = JSON.parse(data);
+		const globalEnv: string[] = json.globalEnv;
+
+		for (const { envVar } of state.entries) {
+			globalEnv.push(envVar);
```

Comment:
I think Omar mentioned that env vars should be sorted alphabetically in this file